_posts/2022-04-24-ui-architecture.md

```diff
@@ -0,0 +1,264 @@
+date:   2022-05-06 08:17:42 -0700
+categories: [rust, gui]
+---
+Rust is an appealing language for building user interfaces for a variety of reasons, especially the promise of delivering both performance and safety. However, finding a good *architecture* is challenging. Architectures that work well in other languages generally don't adapt well to Rust, mostly because they rely on shared mutable state and that is not idiomatic Rust, to put it mildly. It is sometimes asserted for this reason that Rust is a poor fit for UI. I have long believed that it is possible to find an architecture for UI well suited to implementation in Rust, but my previous attempts (including the current [Druid] architecture) have all been flawed. I have studied a range of other Rust UI projects and don't feel that any of those have suitable architecture either.
+
+This post presents a new architecture, which is a synthesis of existing work and a few new ideas. The goals include expression of modern reactive, declarative UI, in components which easily compose, and a high performance implementation. UI code written in this architecture will look very intuitive to those familiar with state of the art toolkits such as SwiftUI, Flutter, and React, while at the same time being idiomatic Rust.
+
+The name "Xilem" is derived from a [xylem], a type of transport tissue in vascular plants, including trees. The word is spelled with an "i" in several languages including Romanian and Malay, and is a reference to [xi-editor], a starting place for explorations into UI in Rust (now on hold).
+
+Like most modern UI architectures, Xilem is based on a *view tree* which is a simple declarative description of the UI. For incremental update, successive versions of the view tree are *diffed,* and the results are applied to a widget tree which is more of a traditional retained-mode UI. Xilem also contains at heart an incremental computation engine with precise change propagation, specialized for UI use.
+
+The most innovative aspect of Xilem is event dispatching based on an *id path,* at each stage providing mutable access to app state. A distinctive feature is Adapt nodes (an evolution of the lensing concept in Druid) which facilitate composition of components. By routing events *through* Adapt nodes, subcomponents have access to a different mutable state reference than the parent.
+
+## Synchronized trees
+
+In each "cycle," the app produces a view treeRendering in Xilem begins with the view tree. This tree has fairly short lifetime; each time the UI is updated, a new tree is generated. From this, a widget tree is built (or rebuilt), and the view tree is retained only long enough to assist in event dispatching and then be diffed against the next version, at which point it is dropped. In addition to these two trees, there is a third tree containing *view state,* which persists across cycles. (The view state serves a very similar function as React hooks)
```

Comment:
The view state is stored as a tree, yes. I'm not familiar with Recoil but I certainly am aware of more general incremental computation engines (Salsa, Adapton, Incremental, etc) in which the dependencies are explicitly modeled as a graph.

This is all potentially a much deeper conversation. I believe, without having gone too deep into implementation, that it would be fairly straightforward to wire up an actual graph incremental engine, and have the view state tree consist of lightweight references into that engine. (such an engine could be made available through context/environment, which is not currently part of the implementation but is planned)

So basically this is an area where I want to see what happens.